### PR TITLE
[JENKINS-72028] replace deprecated `URL()` at  `core/src/main/java/hudson/`

### PR DESCRIPTION
**Pull request by abhishekmaity**
Replace deprecated `URL()` with `URI`#`.toURL()` . Changes implemented for [core/src/main/java/hudson/‎](https://github.com/search?q=repo%3Ajenkinsci%2Fjenkins+%22new+URL%22+lang%3Ajava+path%3A%2F%5Ecore%5C%2Fsrc%5C%2Fmain%5C%2Fjava%5C%2F%2F+path%3A%2F%5Ecore%5C%2Fsrc%5C%2Fmain%5C%2Fjava%5C%2Fhudson%5C%2F%2F&type=code) files only.

See [JENKINS-72028](https://issues.jenkins.io/browse/JENKINS-72028).

### Proposed changelog entries

- JENKINS-72028, replaced deprecated `URL()` with `URI#.toURL()`

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Comment by abhishekmaity** [comment]
requesting @jtnord @timja @NotMyFault

File: core/src/main/java/hudson/tools/ZipExtractionInstaller.java
Comment (jtnord) [inline core/src/main/java/hudson/tools/ZipExtractionInstaller.java:89]: this should not swallow the exception - as the action has not performed this needs to throw an Exception.

I suggest wrapping the URISyntaxException into a MalformedURLException as done elsewhere.

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:3342]: this is ok - but at the same time the class is not public so can only really be used inside the same package (especially since it is not extending anything public - except Object which does not not this method) (and the singleton instance is private). 
Can the callers be adjusted instead?
I only found one caller in a quick scan -> https://github.com/jenkinsci/jenkins/blob/778b8b62de4e6f42da87cd78dbe8d731a6888cf6/core/src/main/java/hudson/FilePath.java#L1013

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java:235]: please try and prevent whitespace changes in unrelated code.

File: core/src/main/java/hudson/cli/InstallPluginCommand.java
Comment (jtnord) [inline core/src/main/java/hudson/cli/InstallPluginCommand.java:108]: neets to be handled in line 117 along with the `MalformedUrlException`

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (jtnord) [inline core/src/main/java/hudson/model/UpdateCenter.java:1214]: do not swallow the exception here.   this already throws `MalformedURLException` (via IOException declaration)

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jtnord) [inline core/src/main/java/hudson/model/UpdateSite.java:222]: please include the cause of the exception like you have elsewhere.

**Review by jtnord — CHANGES_REQUESTED** [review]
there is an occurence where the issue is silently swallowed, also please attempt to keep the diff to a minimum by not reformatting unchanged code.

File: core/src/main/java/hudson/tools/ZipExtractionInstaller.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/tools/ZipExtractionInstaller.java:89]: Implemented the following as per suggestion
```
catch (URISyntaxException e) {
            MalformedURLException mex = new MalformedURLException(e.getMessage());
            mex.initCause(e);
            throw mex;
```

File: core/src/main/java/hudson/FilePath.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/FilePath.java:3342]: I am unable to understand your point here.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/TcpSlaveAgentListener.java:235]: It was produced from `mvn: spotless apply`. Will cross check before commit next time onwards

File: core/src/main/java/hudson/cli/InstallPluginCommand.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/cli/InstallPluginCommand.java:108]: added `e.getMessage();` under `catch(MalformedUrlException e)`

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/model/UpdateCenter.java:1214]: added
```
MalformedURLException mex = new MalformedURLException(e.getMessage());
            mex.initCause(e);
            throw mex;
```

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/model/UpdateSite.java:222]: modified to `throw new MalformedURLException("Incorrect URL data returned");`

**Comment by abhishekmaity** [comment]
@jtnord I have replaced with `e.getCause();` at `core/src/main/java/hudson/cli/InstallPluginCommand.java` now but its getting SpotBugs warning. Can you help to fix it.

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [reply core/src/main/java/hudson/FilePath.java:3342]: > I am unable to understand your point here.

you could confidently let `URISyntaxException` be thrown from this method and adapt the callers as there can be none that are external to this package - and hence there should be done external to this repository.

File: core/src/main/java/hudson/cli/InstallPluginCommand.java
Comment (jtnord) [reply core/src/main/java/hudson/cli/InstallPluginCommand.java:108]: this has the potential to throw `URISyntaxException`,  the code already handled `MalformedURLException` , it is the newly thrown `URISyntaxException` that needs to be handled on line 117 in the same way.

File: core/src/main/java/hudson/cli/InstallPluginCommand.java
Comment (jtnord) [inline core/src/main/java/hudson/cli/InstallPluginCommand.java:119]: You need to handle a potential `URISyntaxException` here - calling `getCause()` is doing mostly nothing.

File: core/src/main/java/hudson/cli/InstallPluginCommand.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/cli/InstallPluginCommand.java:108]: replaced/added with
```
} catch (MalformedURLException | URISyntaxException e) {
                // not an URL
                e.getCause();
            }
```

File: core/src/main/java/hudson/FilePath.java
Comment (jtnord) [inline core/src/main/java/hudson/FilePath.java:1]: I doubt you meant to delete this file.

File: core/src/main/java/hudson/cli/InstallPluginCommand.java
Comment (jtnord) [inline core/src/main/java/hudson/cli/InstallPluginCommand.java:120]: this is dead code - you did not need to add this and spotbugs is probably complaining 

```suggestion
```

**Review by jtnord — CHANGES_REQUESTED** [review]
Seems to have accidently deleted a critical file (`FilePath) and is changing things outside `core/src/main/java/hudson/` (in the PR title and summary).  Looks like a bad merge.

File: core/src/main/java/hudson/FilePath.java
Comment (abhishekmaity) [reply core/src/main/java/hudson/FilePath.java:1]: Sorry, I have added it back

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (jtnord) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java:141]: URI exposes a getHost() so there is there a need to convert to a `URL`?

```suggestion
            return new URI(Jenkins.get().getRootUrl()).getHost();
        } catch (URISyntaxException e) {
```

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jtnord) [inline core/src/main/java/hudson/model/UpdateSite.java:222]:
```suggestion
            MalformedURLException mex = new MalformedURLException(e.getMessage());
            mex.initCause(e);
            throw mex
```

File: core/src/main/java/hudson/model/UpdateSite.java
Comment (jtnord) [inline core/src/main/java/hudson/model/UpdateSite.java:224]:
```suggestion
            throw mex;
```

**Review by jtnord — APPROVED** [review]
assuming the build passes

**Review by NotMyFault — APPROVED** [review]
/label ready-for-merge

---

This PR is now ready for merge. We will merge it after ~24 hours if there is no negative feedback.
Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process.
Thanks!

**Comment by basil** [comment]
Causes [JENKINS-72203](https://issues.jenkins.io/browse/JENKINS-72203). In the absence of a tested fix, I intend to revert this PR toward 2.429.

**Pull request merged**
